### PR TITLE
CockroachDB use "serial" datatypes for version < 21.2, "generated by default" for version >= 21.2

### PR DESCRIPTION
**Pull request by ecapoccia**
## Environment
**Liquibase Version**: 3.8.2

**Liquibase Integration & Version**: <Pick one: CLI, maven, gradle, spring boot, servlet, etc.>

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**:

**Operating System Type & Version**:

## Pull Request Type

* [ ] Bug fix (non-breaking change which fixes an issue.)
* [ ] Enhancement/New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Issue https://github.com/liquibase/liquibase/issues/1862 happens with CDB major version beyond 10.0 (CBD 21.1); this is because it inherits the Postgres settings.
This change is a like for like of the previous behaviour up to CBD 20.2 for instance (returning empty string).

## Steps To Reproduce
List the steps to reproduce the behavior.

* Please be precise and ensure private information is redacted
* Include things like
  * Files used - sql scripts, changelog file(s), property file(s), config files, POM Files
  * Exact commands used - CLI, maven, gradle, spring boot, servlet, etc.

## Actual Behavior
A clear and concise description of what happens in the software **before** this pull request.

* Include console output if relevant
* Include log files if available.

## Expected/Desired Behavior
A clear and concise description of what happens in the software **after** this pull request.

## Fast Track PR Acceptance Checklist:

* [ ] Build is successful and all new and existing tests pass
* [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
* [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

**Comment by kristinelson** [comment]
Thanks @ecapoccia for raising. 

@r2liquibase could we get this PR merged in to fix this issue below? 

This PR fixes this related ticket =https://github.com/liquibase/liquibase/issues/1862

**Comment by r2liquibase** [comment]
Hi, @ecapoccia ! Thanks for the PR! Would you mind looking at the test failures? Thanks!

**Comment by ecapoccia** [comment]
@r2liquibase I've added a commit that fixes the tests and also a few other issues.
Can you trigger another build?

**Review by nvoxland — CHANGES_REQUESTED** [review]
Does #1883 solve the issue for you as well? I think that applies to both postgresql and cockroachdb. It does require setting `liquibase.convertDataTypes=false` in your liquibase.properties file or as a CLI argument because we can't right now change the default behavior for existing users.

**Comment by ecapoccia** [comment]
@nvoxland your PR *might* have the same effect  -- you'd need to test that to verify. 

My point is, the fundamental issue that I'm trying to address is the fact that CDB is not Postgres so there is no reason for CDB to have a different behaviour if the major version of it is higher or lower than 10.0.

It's not intuitive why one would want to set liquibase.convertDataTypes = false to fix a different behaviour of CDB between v 9.0 and v 13.0.

Side note. I believe in your PR you possibly missed the SmallInt data type which is similar to the other two.

**Comment by nvoxland** [comment]
I sync'ed master into this PR and also updated it a bit.

I see how the connection's getMajorVersion() call is returning "10" for cockroach 21.1 vs "9" for 20.1, and neither one is right.

Looking at the cockroach docs, it looks like they now support the GENERATED BY DEFAULT AS IDENTITY preferred syntax for 21.2+, so I overrode the getMajorVersion() logic to get the actual cockroach versions and then made a new useSerialDatatype() function which any postgresql-like databases can control as they need.

**Review by nvoxland — APPROVED** [review]

**Comment by nvoxland** [comment]
Code review and test results:

Things to be aware of
- Refactored logic used by both postgresql and cockroachdb
- Cockroachdb 21.2 adds "identity columns" so we use `generated by default` logic on that version and `serial` for earlier versions https://www.cockroachlabs.com/docs/stable/create-table.html#identity-columns
- This change fixed a lot of the failures that were keeping me from enabling CockroachIntegrationTest from running on the build, but there was still one unrelated failure. But, the automation is much happier with this change.
- I tested update logic for 20.1, 21.1, and 21.2
- Automated tests for different sql for different versions will come along with the in-progress integration test frameworks
 
Things to worry about
- Nothing

**Review by suryaaki2 — APPROVED** [review]

File: liquibase-core/src/test/groovy/liquibase/database/core/CockroachDatabaseTest.groovy [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package liquibase.database.core
```
Comment (XDelphiGrl) [inline liquibase-core/src/test/groovy/liquibase/database/core/CockroachDatabaseTest.groovy:1]: @nvoxland, thank you for adding this excellent set of tests. Do these tests run during the build? I don't see CRDB in the list of checks, but perhaps CRDB tests run in the Postgres bucket?

**Comment by XDelphiGrl** [comment]
This PR addresses an issue related to retrieving the Cockroach database version. As part of this fix, Liquibase generates version-specific autoincrement SQL statements for Cockroach identity columns. 

#### CRDB Versions Less than 21.2
* Generated SQL for autoincrement SQL statements use SMALLSERIAL, SERIAL and BIGSERIAL. For example,

```
CREATE TABLE public.pr1908 (id_smallint SMALLSERIAL, id_int SERIAL, id_integer SERIAL, id_int8 BIGSERIAL, id_int64 INT64, id_bigint BIGSERIAL);
```

#### CRDB Versions Greater than or Equal to 21.2
* Generated SQL for autoincrement SQL statements use GENERATED BY DEFAULT AS IDENTITY. For example,

```
CREATE TABLE public.pr1908 (id_smallint SMALLINT GENERATED BY DEFAULT AS IDENTITY, id_int INTEGER GENERATED BY DEFAULT AS IDENTITY, id_integer INTEGER GENERATED BY DEFAULT AS IDENTITY, id_int8 BIGINT GENERATED BY DEFAULT AS IDENTITY, id_int64 INT64 GENERATED BY DEFAULT AS IDENTITY, id_bigint BIGINT GENERATED BY DEFAULT AS IDENTITY);-->
```

--- 
Tests added to liquibase/liquibase-test-harness to validate changes; PR pending. 

---
**Test Environment**
Liquibase Core: master/1650/1efd92, Pro: master/658/91c564
* This is the master branch of fork liquibase/ecapoccia, not master of liquibase/liquibase

Cockroach 20.1, 20.2, 21.1
Postgres 9, 9.5, 10, 11, 12, 13, 14
[Passing Functional Tests](https://github.com/liquibase/liquibase-pro-tests/actions/runs/1917218154)
[Liquibase Test Harness PR](https://github.com/liquibase/liquibase-test-harness/pull/191)

**Pull request merged**